### DISCUSSION TITLE
How to weave aspects using AspectjCompile task

### DISCUSSION THREAD

**Discussion by libertywork-agb** [freefair/gradle-plugins#493] [Q&A]
I am using the `io.freefair.aspectj.base` plugin version 6.3.0. Referencing the [documentation](https://docs.freefair.io/gradle-plugins/6.3.0/reference/#_aspectjcompile), it can be used in this way:
```groovy
task myAjcTask(type: io.freefair.gradle.plugins.aspectj.AspectjCompile) {
    aspectjClasspath.setFrom configurations.aspectj
    ajcOptions {
        inpath = files()
        aspectpath = files()
    }
}
```
I created an `aspectj` configuration (it did not exist from applying the plugin), declared a dependency to the subproject containing my aspects using that configuration, defined my custom ajc task, and finalized `compileJava` by my ajc task.
```groovy
plugins {
    id 'java'
    id 'io.freefair.aspectj.base' version '6.3.0'
}

configurations {
    aspectj
}

dependencies {
    aspectj project(':my-aspect-lib')
}

task myAjcTask(type: io.freefair.gradle.plugins.aspectj.AspectjCompile) {
    aspectjClasspath.setFrom configurations.aspectj
    ajcOptions {
        inpath.setFrom compileJava.outputs.files
        aspectpath.setFrom configurations.aspectj
    }
}

compileJava.finalizedBy tasks.myAjcTask
```
This setup results in the following exception when running `compileJava`:
```
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':html-sanitizer-test:myAjcTask'.
Caused by: java.lang.NullPointerException
	at io.freefair.gradle.plugins.aspectj.internal.AspectJRuntime.findAspectjtoolsJarFile(AspectJRuntime.java:108)
	at io.freefair.gradle.plugins.aspectj.internal.AspectJRuntime.access$000(AspectJRuntime.java:28)
	at io.freefair.gradle.plugins.aspectj.internal.AspectJRuntime$1.inferAspectjClasspath(AspectJRuntime.java:67)
	at io.freefair.gradle.plugins.aspectj.internal.AspectJRuntime$1.createDelegate(AspectJRuntime.java:59)
```
`aspectjClasspath` does not have much in the way of documentation, so I don't really understand its importance, but it seems like I have it set correctly based on the example `myAjcTask` from the documentation. My problem is that I don't know exactly what to put for `inpath` and `aspectpath`.

From my understanding, `inpath` should be the files into which aspects are woven and `aspectpath` should be the files containing the aspects to weave. I want the compiled Java to be woven with `my-aspect-lib` aspects.

If I just don't set `inpath` or if I set it to an empty file collection (`files()`) then running the build succeeds but of course it doesn't weave any aspects.

I also tried upgrading to version 6.4.1 of the plugin and got the exact same results.

Why am I getting this exception and how should `inpath` and `aspectpath` be configured?

**Comment by larsgrefer** [comment]
You are basically doing, what the `io.freefair.aspectj.post-compile-weaving` does.

> From my understanding, `inpath` should be the files into which aspects are woven and `aspectpath` should be the files containing the aspects to weave. I want the compiled Java to be woven with my-aspect-lib aspects.

That is correct.
`aspectjClasspath` should contain the aspectj compiler itself.

The `io.freefair.aspectj.base` plugin tries to infer the ajc version based on the classpath of the corresponding compile task, but this fails here because the compile classpath is null.

**Reply by libertywork-agb** [reply]
I understand that `io.freefair.aspectj.post-compile-weaving` does basically the same thing, but in my [other discussion Q&A post](https://github.com/freefair/gradle-plugins/discussions/471) you indicated that it was not possible to exclude specific project classes from being processed. I need a Gradle plugin that can weave only certain aspects and exclude project aspects.

I'm afraid I don't quite understand what you mean by compile classpath.

If I add these lines to my task:
```groovy
println 'compileJava.outputs.files: ' + compileJava.outputs.files
compileJava.outputs.files.forEach(file -> println file)
```

I get this output:
```
compileJava.outputs.files: task 'compileJava' output files
/home/user/projects/root-project/consuming-subproject/build/classes/java/main
/home/user/projects/root-project/consuming-subproject/build/generated/sources/annotationProcessor/java/main
/home/user/projects/root-project/consuming-subproject/build/generated/sources/headers/java/main
/home/user/projects/root-project/consuming-subproject/build/tmp/compileJava/previous-compilation-data.bin
```

If I add these lines to my task:
```groovy
println 'configurations.aspectj: ' + configurations.aspectj
configurations.aspectj.forEach(file -> println file)
```

I get this output:
```
configurations.aspectj: configuration ':consuming-subproject:aspectj'
/home/user/projects/root-project/my-aspect-lib/build/libs/my-aspect-lib-0.0.1.jar
/home/user/.gradle/caches/modules-2/files-2.1/org.aspectj/aspectjrt/1.9.7/ha5h/aspectjrt-1.9.7.jar
...
```

If I add any of the following dependencies I still get the same exception:
```groovy
compileClasspath 'org.aspectj:aspectjrt:1.9.7'
compileClasspath 'org.aspectj:aspectjweaver:1.9.7'
compileClasspath 'org.aspectj:aspectjtools:1.9.7'
```

How can the compile classpath be set, and how can it be ensured that the AspectJ classpath contains the AspectJ compiler?

**Reply by larsgrefer** [reply]
I've made some changes to the plugin which should make your use-case a bit easier (and the error message less confusing).

The `io.freefair.gradle.plugins.aspectj.AspectjCompile` task has four properties that are relevant here:

- `aspectjClasspath`: This file collection must contain the `aspectjtools.jar` and will be used to actually invoke the `ajc` compiler.
- `classpath`: This file collection will be directly mapped to the `-classpath` argument of `ajc`. When using the `io.freefair.aspectj.base` plugin, this defaults to null. When using the other plugins it should be correctly wired to use the `compileClasspath` configuration.
- `ajcOptions.inpath`: This file collection will be mapped to the `-inpath` argument of `ajc`
- `ajcOptions.aspectpath`: This file collection will be mapped to the `-aspectpath` argument of `ajc`.

More information about the `ajc` arguments can be found here: https://www.eclipse.org/aspectj/doc/released/devguide/ajc-ref.html

In its current form, the base plugin tries to determine the `aspectjtools.jar` for the `aspectjClasspath` by inspecting the `classpath`.
For the other plugins this works well, but in your case it failed, because `classpath` was null.

With my changes, which will be available with the next release, the plugin will inspect both the `classpath` and the `ajcOptions.aspectpath` to find the correct `ajc` version. This should work for you, since your aspectpath contains a aspectjrt.jar
